std/numeric.d

```diff
@@ -1400,6 +1400,227 @@ unittest
+    relTolerance = Relative tolerance.
+    absTolerance = Absolute tolerance.
+
+Preconditions:
+    $(D ax) and $(D bx) shall be finite reals. $(BR)
+    $(D relTolerance) shall be normal positive real. $(BR)
+    $(D absTolerance) shall be normal positive real no less then $(D T.epsilon*2).
+
+Returns:
+    A tuple consisting of $(D x), $(D y = f(x)) and $(D error = 3 * (absTolerance * fabs(x) + relTolerance)).
+
+    The method used is a combination of golden section search and
+successive parabolic interpolation. Convergence is never much slower
+than that for a Fibonacci search.
+
+References:
+    "Algorithms for Minimization without Derivatives", Richard Brent, Prentice-Hall, Inc. (1973)
+
+See_Also: $(LREF findRoot), $(XREF math, isNormal)
++/
+Tuple!(T, "x", Unqual!(ReturnType!DF), "y", T, "error")
+findLocalMin(T, DF)(
+        scope DF f,
+        in T ax,
+        in T bx,
+        in T relTolerance = sqrt(T.epsilon),
+        in T absTolerance = sqrt(T.epsilon),
+        )
+    if(isFloatingPoint!T)
+in
+{
+    assert(isFinite(ax), "ax is not finite");
+    assert(isFinite(bx), "bx is not finite");
+    assert(isNormal(relTolerance) && relTolerance > 0,
+        "relTolerance shall be normal positive real");
+    assert(isNormal(absTolerance) && absTolerance >= T.epsilon*2,
+        "absTolerance shall be normal positive real no less then $(D T.epsilon*2)");
+}
+out(result)
+{
+    assert(isFinite(result.x));
+}
+body
+{
+    alias R = Unqual!(CommonType!(ReturnType!DF, T));
+    // c is the squared inverse of the golden ratio
+    // (3 - sqrt(5))/2
+    // Value obtained from Wolfram Alpha.
+    immutable T c = 0x0.61c8864680b583ea0c633f9fa31237p+0L;
+    immutable T cm1 = 0x0.9e3779b97f4a7c15f39cc0605cedc8p+0L;
+    R tolerance;
+    T a = ax > bx ? bx : ax;
+    T b = ax > bx ? ax : bx;
+    //sequence of declarations suitable for SIMD instructions
+    T  v = a * cm1 + b * c;
+    assert(isFinite(v));
+    R fv = f(v);
+    //R fv = f(v);
+    if(isNaN(fv) || fv == -T.infinity)
+        return typeof(return)(v, fv, T.init);
+    T  w = v;
+    R fw = fv;
+    T  x = v;
+    R fx = fv;
+    for(R d = 0, e = 0;;)
+    {
+        T m = (a + b) / 2;
+        // This fix is not part of the original algorithm
+        if(!isFinite(m)) // fix infinity loop. Issue can be reproduced in R.
+        {
+            m = a / 2 + b / 2;
+            if(!isFinite(m)) // fast-math compiler switch is enabled
+            {
+                //SIMD instructions can be used by compiler, do not reduce declarations
+                int a_exp = void;
+                int b_exp = void;
+                immutable an = frexp(a, a_exp);
+                immutable bn = frexp(b, b_exp);
+                immutable am = ldexp(an, a_exp-1);
+                immutable bm = ldexp(bn, b_exp-1);
+                m = am + bm;
+                if(!isFinite(m)) // wrong input: constraints is disabled in release mode
```

Comment:
If I understand the comment correctly, this is to break out of a loop caused by bad user input, which is not caught because `assert` gets disabled in release mode. Is this correct? If so, wouldn't it be better to *always* test the input, e.g. with `enforce`?